### ISSUE TITLE
Some questions about tensorforce

### ISSUE THREAD

**Issue by marooncn** [tensorforce/tensorforce#565]
Hi, thanks for your great work. But when I read the docs I have some questions about this framework.

Q1: How does the network update? Is it  `agent.observe(terminal=terminal, reward=reward)` collect gradients until the specified _timesteps/episodes_ in _update_model_?

Q2: Does the output layer of network define automatically when we define an agent? Such as I define an
_DQNAgent_ which has three actions to choose and I don't need to define the last layer is `dict(type='dense', size=3, activation='softmax')`.

Q3: _DQNAgent_ needs to collect _[St, a, r, St+1]_, in the following examples:
```
while True:
    state2=f(state)  
    action = agent.act(states=state2)
    action2=g(action) 
    state, reward, terminal = environment.execute(actions=action2)
    agent.observe(reward=reward, terminal=terminal)
```
Does it collects [state2, action2, r, state2'] or [state, action, r, state'] ?

Q4: How can I output training loss ?

Actually, I use a DQNAgent to realize robot navigation task. the input is compressed image, goal and the previous action. The output is three actions(forwrd, left, right) to choose. The agent is defined as:
```
network_spec = [
    dict(type='dense', size=128, activation='relu'),
    dict(type='dense', size=32, activation='relu')
]

memory = dict(
    type='replay',
    include_next_states=True,
    capacity=10000
)

exploration = dict(
    type='epsilon_decay',
    initial_epsilon=1.0,
    final_epsilon=0.1,
    timesteps=10000,
    start_timestep=0
)

update_model = dict(
    unit='timesteps',
    batch_size=64,
    frequency=64
)

optimizer = dict(
    type='adam',
    learning_rate=0.0001
)

agent = DQNAgent(
    states=dict(shape=(36,), type='float'), 
    actions=dict(shape=(3,), type='int'), 
    network=network_spec,
    update_mode=update_model,
    memory=memory,
    actions_exploration=exploration,
    optimizer=optimizer,
    double_q_model=True
)
```
Because I need to deal with the captured image to a compressed vector as a part of state, I run an episode as the following rather than using _runner_.
```
    while True:
        compressed_image = compress_image(observation)   # map the capture image to a 32-dim vector
        goal = env.goal   # shape(2, )
        pre_action = action  # shape(2, )
        state = compressed_image + goal + pre_action
        action = agent.act(state)
        observation, terminal, reward = env.execute(action)
        agent.observe(terminal=terminal, reward=reward)
        timestep += 1
        episode_reward += reward
        if terminal or timestep == max_timesteps:
            success = env.success
            break
```
Can it work? I haved trained much time but the result is not ideal. So I want to know if I use tensorforce correctly. Thank you!

**Comment by AlexKuhnle** [comment]
Hi @marooncn,
Q1: Yes, updates are implicitly performed as part of `observe()` as specified by `update_mode`.
Q2: Yes, a final linear layer is automatically added for handling the action(-distribution)-specific outputs.
Q3: The first version, with `state2` etc, is correct, as the other variables are never passed to the agent.
Q4: Your script looks alright, only thing that may be missing is `agent.reset()` before the while loop. Getting the agent to properly learn is unfortunately not always straightforward, so you may want to try how changing hyperparameters affects the result. Also, maybe try PPO instead, as it seems to generally work best (at least without extensive tuning).

**Comment by marooncn** [comment]
@AlexKuhnle , Thanks for your response. I find something abnormal that when I print the output of my agent, its value changes a little. For a test, you can run the following code:
```
from tensorforce.agents import PPOAgent
import numpy as np
 
network_spec = [
     dict(type='dense', size=128, activation='relu'),
     dict(type='tf_layer', layer='batch_normalization'),
     dict(type='dense', size=32, activation='relu'),
     dict(type='tf_layer', layer='batch_normalization')]

exploration=dict(
     type='epsilon_decay',
     initial_epsilon=1.0,
     final_epsilon=0.1,
     timesteps=100000,
     start_timestep=0
)

optimizer = dict(
     type='adam',
     learning_rate=0.001
)

agent = PPOAgent(
     states=dict(shape=(2,), type='float'),  # GazeboMaze.states,
     actions=dict(shape=(2,), min_value=0, max_value=1, type='float'),
     network=network_spec,
     # memory=memory,
     actions_exploration=exploration,
     step_optimizer=optimizer
)

while True:
     state = np.random.choice(10, 2)
     action = agent.act(state)
     terminal, reward = [False, 1]
     agent.observe(terminal=terminal, reward=reward)
     print(action)
```
The input is a 2-dim vector randomly chosen from [0, 10], but when we run this code the output is always [1, 1] even I have added BN after each network layer. What I want is that it can output the value evenly distributed in the specified range to explore efficiently. I guess it's because of the automatic-defined output layer, I want to add BN to the output layer and define the different initialization function of network weight but I don't find the related API. Can you help me to check it?

**Comment by AlexKuhnle** [comment]
Interesting... it does vary more later on, but initially it is indeed always/mostly 1.0, which is not really what should happen. I'll have a look...

**Comment by AlexKuhnle** [comment]
Found the problem, actually a bug in the continuous exploration calculation -- thanks for reporting! See if it works better with the latest commit.

**Comment by marooncn** [comment]
Hi, @AlexKuhnle . I find it's because exploration in the above test code, if I use the default exploration function it will be OK. But for my navigation agent, its output still changes a little. Maybe I need use _tf_apply_ to define the output layer by myself.

**Comment by AlexKuhnle** [comment]
Does it work better with the new version? At least it shouldn't be always 1.0...

**Comment by marooncn** [comment]
@AlexKuhnle , I update tensorforce and it looks better when run the test code. But I think it still many limit value(-1 or 1) generated: 

> [1.         0.45494115]
> [ 0.24685547 -0.91656864]
> [0.61867374 0.15192983]
> [0.4791621 1.       ]
> [-0.17667249 -0.65035665]
> [ 0.08686978 -0.03326213]
> [-1.         -0.38841623]
> [1.        0.9778045]
> [ 1.         -0.33899912]
> [-0.26226828 -0.2862496 ]
> [-0.29749477 -1.        ]
> [ 1.         -0.24625593]
> [-1.         -0.31478393]
> [-0.3683312  -0.14433295]
> [0.77993083 0.16400331]
> [-0.25849098 -0.56925845]
> [-0.89845586 -1.        ]
> [ 1.        -0.8586832]
> [-1.          0.69175935]

And I use tf_apply to define the network as the following according to this [file](https://github.com/tensorforce/tensorforce/blob/38016f4b20078d1daf02e7b5e9359547d55adfc7/tensorforce/tests/test_ddpg_agent_custom_critic.py):
```
class DQNetwork(Network):
    def tf_apply(self, x, internals, update, return_internals=False):
        import tensorflow as tf
        # fc1
        fc1 = tf.layers.dense(x, 128, kernel_initializer=tf.contrib.layers.xavier_initializer(), activation=None)
        fc1 = tf.layers.batch_normalization(fc1, training=restore)
        fc1 = tf.nn.relu(fc1)
        # fc2
        fc2 = tf.layers.dense(fc1, 32, kernel_initializer=tf.contrib.layers.xavier_initializer(), activation=None)
        fc2 = tf.layers.batch_normalization(fc2, training=restore)
        fc2 = tf.nn.relu(fc2)
        # output
        output = tf.layers.dense(fc2, 3, kernel_initializer=tf.contrib.layers.xavier_initializer(), activation=None)
        output = tf.layers.batch_normalization(output, training=restore)
        output = tf.nn.softmax(output)

        if return_internals:
            return output, None
        else:
            return output

agent = DQNAgent(
    states=dict(shape=(36,), type='float'), 
    actions=dict(shape=(3,), type='int'), 
    network=DQNetwork,
)
```
But there is something wrong:

>   File "/home/maroon/tensorforce/tensorforce/models/model.py", line 589, in setup_components_and_tf_funcs
>     custom_getter_=custom_getter

**Comment by AlexKuhnle** [comment]
I would recommend using the network configuration you posted above, as otherwise you have to take care of variable handling yourself, which can be problematic with frameworks like `tf.layers`. Also, note that the output layer will nonetheless be added automatically, it's independent of the network and would require changing distributions (which, again, is not really something I would recommend).

**Comment by AlexKuhnle** [comment]
What is it you want to achieve that makes you want to use your own config?

**Comment by AlexKuhnle** [comment]
Oh, and why it produces many limit values: because the noise is gaussian and clipped at the limits... maybe not great, but not sure about a good way to fix that.

**Comment by AlexKuhnle** [comment]
(Reducing the exploration value maybe -- note that this value gives the stddev of the noise added, so may make sense to not start with 1.0)

**Comment by marooncn** [comment]
Hi @AlexKuhnle , in the [blog](https://reinforce.io/blog/introduction-to-tensorforce/), the network builder function can be  programmatically defined for some special cases, such as the handling multiple input states. And I define it according to it, but it can't work.

**Comment by marooncn** [comment]
> Oh, and why it produces many limit values: because the noise is gaussian and clipped at the limits... maybe not great, but not sure about a good way to fix that.
And the output activation function should be softmax for DQNAgent and it should scale the output to the defined range. Why does it need to clip to limit?

**Comment by AlexKuhnle** [comment]
You're right. Actually, there is a good way of handling multiple states now, which I can tell you about. But anyway, what I'm not sure about is if you use high-level APIs like `tf.layers`, which handle variables internally -- compare with the custom network you posted, which uses more low-level tf primitives and explicit variable allocation.

My comment about clipping referred to an earlier config, where `actions=dict(shape=(2,), min_value=0, max_value=1, type='float')`, which means continuous actions using a Beta distribution and clipped exploration.

Regarding softmax, Tensorforce always, to any network, adds a final action layer with linear transformation and appropriate processing (like softmax or so), so while you can in principle use the softmax as you showed, you likely have in mind that this is the final layer processing, in which case you can just remove this part, as it is taken care of internally.

Documentation is the problem here, as so often -- sorry for it being insufficient. I hope to improve that after the new version is finished. I think it may make sense to already switch to the new version, if you want to give it a go (not too many external changes, but some things work better). Also, how about taking this conversation to Gitter, happy to be of help figuring things out.

**Comment by marooncn** [comment]
So what if I want to handle multiple states. Such as the following case:
![default](https://user-images.githubusercontent.com/29606478/53557571-29d29f80-3b81-11e9-8db6-fcffcda8faab.png)
How could I define this network for the DQNAgent?

**Comment by AlexKuhnle** [comment]
[This code](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/tests/base_agent_test.py#L134) in the unittests may be of interest to you. Basically, you define your network as a list of lists, where each block defines a layer-stack with a first `"input"` layer retrieving one or multiple states (or previously defined outputs) and a last `"output"` layer registering outputs for later retrieval. Note the `aggregation_type` argument for `"input"` in case of multiple inputs.

**Comment by marooncn** [comment]
@AlexKuhnle Thank you, the network is defined as a list of lists as you said and it can work now.
But I still have a question about the output. Can I close the output layer which the agent automatic generated (For DQNAgent it's necessary to automatically  generate  the output layer but for other agent such as PPOAgent I think the output layer should be defined by user to choose the reasonable activation function)? Such as I want to generate action in the range [0, 1], then I can define the last layer activation
as 'sigmoid'  But actually the output varies out of the limits. For a test:
```
from tensorforce.agents import PPOAgent
import numpy as np
network_spec = [
     [
          dict(type='input', names=['state']),
          dict(type='dense', size=32, activation='relu'),
          dict(type='output', name='image_output'),
     ],

     [
          dict(type='input', names=['image_output', 'action'], aggregation_type='concat'),
          dict(type='dense', size=2, activation='sigmoid'),
     ]
]

agent = PPOAgent(
     states=dict(
          state=dict(shape=(2,), type='float'),
          action=dict(shape=(2,), type='float')
     ),
     actions=dict(shape=(2,), type='float'),
     network=network_spec,
)

while True:
     state = np.random.randn(2)
     action = np.random.randn(2)
     states = dict(state=state, action=action)
     action = agent.act(states)
     terminal, reward = [False, 1]
     agent.observe(terminal=terminal, reward=reward)
     print(action)
```

**Comment by AlexKuhnle** [comment]
It's actually the case that for DQN it would be easier to leave the output layer to the user, while for PPO the network parametrizes a distribution, from which actions are sampled and whose log probability is used for the update, so it's not easily possible to specify activations as you do.

You can specify `min/max_value` as you did above (`actions=dict(shape=(2,), min_value=0, max_value=1, type='float'`), so in this case you would specify:
```python
actions=dict(
    action1=dict(shape=(), type='float', min_value=-1.0, max_value=1.0),
    action2=dict(shape=(), type='float', min_value=0.0, max_value=1.0)
)
```

**Comment by marooncn** [comment]
@AlexKuhnle , it works, thank you.

**Comment by AlexKuhnle** [comment]
Great! :-)

**Comment by marooncn** [comment]
Hi @AlexKuhnle. [Our work](https://github.com/marooncn/navbot) is completed and open source. The RL framework is tensorforce. Just show the happiness with you. Thanks for your help and your great work.

**Comment by AlexKuhnle** [comment]
Hey @marooncn, that looks really cool, glad to hear you got it all to work. :-)  Thanks for letting me know. I will have a closer look at it later, and may be in touch (probably via Gitter) regarding 1-2 things, if you don't mind.

**Comment by marooncn** [comment]
@AlexKuhnle No problem.
